### ISSUE TITLE
kvserver: failure while transferring lease away when exiting joint config can lead to range unavailability

### ISSUE THREAD

**Issue by arulajmani** [cockroachdb/cockroach#83687]
**Describe the problem**

@nvanbenschoten, @andreimatei, @shralex, @knz and I were looking into a closed loop workload that had been stalled for around 10 minutes. We correlated the stall with an uptick in backoffs due to NLHE errors in the dist sender. 

<img width="266" alt="Screen Shot 2022-06-30 at 9 06 37 PM" src="https://user-images.githubusercontent.com/6366031/176802734-0a72e87a-da03-4368-8556-e09a8c104606.png">

Digging a bit deeper in the logs for one of the nodes, it turned out all these errors pertained to a single range `r54` and all NLHE errors indicated that `n16` was the lease holder. Interestingly, `n16`s replica was of type `VOTER_DEMOTING`. We also see `n4` has a `VOTER_INCOMING` replica, so the range in question is undergoing a lateral lease transfer using a joint config. 

```
...
E220630 23:39:00.154796 2779303 kv/kvclient/kvcoord/dist_sender.go:2268 ⋮ [n1,client=10.11.28.172:35106,user=root] 109206 application error: ‹[NotLeaseHolderError] lease held by different store; r54: replica (n4,s4):17VOTER_INCOMING not lease holder; current lease is repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0›: "sql txn" meta={id=3e7719c8 key=‹/Table/106/1/-9207304179336853575/0› pri=0.00128420 epo=0 ts=1656629806.010494656,0 min=1656629806.010494656,0 seq=0} lock=true stat=PENDING rts=1656629806.010494656,0 wto=false gul=1656629806.510494656,0
I220630 23:39:00.154970 2779303 kv/kvclient/kvcoord/dist_sender.go:1659 ⋮ [n1,client=10.11.28.172:35106,user=root] 109207 evicting range desc ‹desc:r54:/Table/106{-/1/-9131597190716917760} [(n9,s9):13, (n5,s5):2, (n16,s16):16VOTER_DEMOTING_LEARNER, (n20,s20):8NON_VOTER, (n31,s31):14NON_VOTER, (n38,s38):11NON_VOTER, (n4,s4):17VOTER_INCOMING, next=18, gen=58] lease:<nil> spec desc: <nil>› after ‹failed to send RPC: sending to all replicas failed; last error: [NotLeaseHolderError] lease held by different store; r54: replica (n4,s4):17VOTER_INCOMING not lease holder; current lease is repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0›
E220630 23:39:00.178641 2778510 kv/kvclient/kvcoord/dist_sender.go:2268 ⋮ [n1] 109211 application error: ‹[NotLeaseHolderError] lease held by different store; r54: replica (n9,s9):13 not lease holder; current lease is repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0›: "sql txn" meta={id=fc837e32 key=‹/Table/106/1/-9139405485398707370/0› pri=0.00621197 epo=0 ts=1656629805.744869361,0 min=1656629805.744869361,0 seq=1501} lock=true stat=STAGING rts=1656629805.744869361,0 wto=false gul=1656629806.244869361,0 int=500 ifw=1499
E220630 23:39:00.181837 2778510 kv/kvclient/kvcoord/dist_sender.go:2157 ⋮ [n1] 109214 ‹[NotLeaseHolderError] lease held by different store; r54: replica (n5,s5):2 not lease holder; current lease is repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0›: "sql txn" meta={id=fc837e32 key=‹/Table/106/1/-9139405485398707370/0› pri=0.00621197 epo=0 ts=1656629805.744869361,0 min=1656629805.744869361,0 seq=1501} lock=true stat=STAGING rts=1656629805.744869361,0 wto=false gul=1656629806.244869361,0 int=500 ifw=1499
E220630 23:39:00.409114 2778510 kv/kvclient/kvcoord/dist_sender.go:2268 ⋮ [n1] 109359 application error: ‹[NotLeaseHolderError] lease held by different store; r54: replica (n5,s5):2 not lease holder; current lease is repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0›: "sql txn" meta={id=fc837e32 key=‹/Table/106/1/-9139405485398707370/0› pri=0.00621197 epo=0 ts=1656629805.744869361,0 min=1656629805.744869361,0 seq=1501} lock=true stat=STAGING rts=1656629805.744869361,0 wto=false gul=1656629806.244869361,0 int=500 ifw=1499
I220630 23:39:00.409345 2778510 kv/kvclient/kvcoord/dist_sender.go:1659 ⋮ [n1] 109360 evicting range desc ‹desc:r54:/Table/106{-/1/-9131597190716917760} [(n9,s9):13, (n5,s5):2, (n16,s16):16VOTER_DEMOTING_LEARNER, (n20,s20):8NON_VOTER, (n31,s31):14NON_VOTER, (n38,s38):11NON_VOTER, (n4,s4):17VOTER_INCOMING, next=18, gen=58] lease:repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0 spec desc: <nil>› after ‹failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r54: replica (n5,s5):2 not lease holder; current lease is repl=(n16,s16):16VOTER_INCOMING seq=8 start=1656629768.388660032,0 epo=1 pro=1656629767.668815515,0›
```

Looking at `n16s` logs next, we find log lines of the form: 

```
I220630 23:45:43.994460 5373681 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 241 request range lease (attempt #2440)
I220630 23:45:43.994472 5373167 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 242 request range lease (attempt #2886)
I220630 23:45:43.994483 3345317 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 243 request range lease (attempt #3554)
I220630 23:45:43.994497 5373402 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 244 request range lease (attempt #2681)
I220630 23:45:43.994509 5363466 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 245 request range lease (attempt #1342)
I220630 23:45:43.994526 5373472 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 246 request range lease (attempt #2615)
I220630 23:45:43.994543 5374278 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 247 request range lease (attempt #1956)
I220630 23:45:43.994560 5374712 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 248 request range lease (attempt #1608)
I220630 23:45:43.994572 5372414 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 249 request range lease (attempt #3484)
I220630 23:45:43.994589 5376030 kv/kvserver/replica_range_lease.go:1329 ⋮ [n16,s16,r54/16:‹/Table/106{-/1/-913…}›] 250 request 
```

We also noticed from the ranges status page that `n16`'s status is marked as `PROSCRIBED`. This checks out, given it appears as the leaseholder to all other nodes in the cluster yet it is repeatedly trying to acquire the lease for `r54`.  It is repeatedly failing to acquire the lease however, which seems to be running into this check which only allows `VOTER_FULL` and `VOTER_INCOMING` replicas to acquire a lease. 

https://github.com/cockroachdb/cockroach/blob/eeb72360fe8e5bdc9ed123612c4e71b975b52124/pkg/roachpb/metadata_replicas.go#L527-L531

The proposed solution here would be to as simple as relaxing the check about to allow a `VOTER_DEMOTING` replica to acquire the lease if it is part of a joint configuration with another `VOTER_INCOMING` replica. 

Jira issue: CRDB-17207

**Comment by nvanbenschoten** [comment]
I think this issue may be exacerbated by https://github.com/cockroachdb/cockroach/commit/034611b5a4e9616d8dbce47379542786fe1adf16. That commit adds a new case where lease transfers can fail after the outgoing lease has been revoked (placed in a PROSCRIBED state). We see in https://github.com/cockroachdb/cockroach/issues/83261 that we now hit this case more often. I suspect that this is why the lease ends up in a PROSCRIBED state in this issue.

That doesn't mean that we shouldn't urgently fix this issue, but it may be more rare on v22.1 (which does not have 034611b5a4e9616d8dbce47379542786fe1adf16) than we originally thought.

**Comment by arulajmani** [comment]
> I think this issue may be exacerbated by https://github.com/cockroachdb/cockroach/commit/034611b5a4e9616d8dbce47379542786fe1adf16.

Anecdotally, I started seeing this only after a rebase and once I saw it the first time I kept on hitting it quite frequently.
